### PR TITLE
Fix allocate to allocate correctly in all cases

### PR DESCRIPTION
**Pull request by charleso**
Gah! :(

### CODE REVIEW

File: mundane-bytes/src/main/scala/com/ambiata/mundane/bytes/Buffer.scala [hunk 1]
```diff
@@ -65,17 +65,18 @@ object Buffer {
     wrapArray(b.bytes, b.offset + o, b.length - o)
   }
 
-  def reset(b: Buffer): Unit =
+  def reset(b: Buffer): Buffer = {
```
Comment (charleso) [inline mundane-bytes/src/main/scala/com/ambiata/mundane/bytes/Buffer.scala]: I was thinking about this some more and decided it was the wrong behaviour. Happy to drop it if you prefer the old version.

File: mundane-bytes/src/main/scala/com/ambiata/mundane/bytes/Buffer.scala [hunk 2]
```diff
@@ -65,17 +65,18 @@ object Buffer {
...
     b._length = 0
+    b
+  }
 
   /** Ensure there is enough space in the array, and if not return a _new_ [[Buffer]] and array with the original values */
   def allocate(b1: Buffer, i: Int): Buffer =
-    if (b1.length + i <= b1.bytes.length) {
+    if (b1.length + b1.offset + i <= b1.bytes.length) {
```
Comment (charleso) [inline mundane-bytes/src/main/scala/com/ambiata/mundane/bytes/Buffer.scala]: :(

Comment (markhibberd) [inline mundane-bytes/src/main/scala/com/ambiata/mundane/bytes/Buffer.scala]: Still don't know if it this is correct, I think I may of glossed over the sematics of this when I read it last time - but I expected `i` to be a new length, as in you are ensuring that the buffer is big enough for `offset + length`.

File: mundane-bytes/src/main/scala/com/ambiata/mundane/bytes/Buffer.scala [hunk 3]
```diff
@@ -65,17 +65,18 @@ object Buffer {
...
       b1._length += i
```
Comment (markhibberd) [inline mundane-bytes/src/main/scala/com/ambiata/mundane/bytes/Buffer.scala]: Yeh, I missed this totally, I would of really expected `_length = i`

**Pull request merged**